### ISSUE TITLE
Improve versioning with Git tags / GH releases.

### ISSUE THREAD

**Issue by StarpTech** [mxschmitt/playwright-go#59]
Hello,
thanks for the lib It works great! I faced with some limitations:

- I would like to avoid downloading https://github.com/mxschmitt/playwright-go/blob/c9d1531e0354a7f267237b3ac55595a8f18170e4/run.go#L29 after every container start. I assumed that playwright is already installed on the official `mcr.microsoft.com/playwright:bionic` images.

- It looks like that playwright is installed from an external source https://github.com/mxschmitt/playwright-go/blob/c9d1531e0354a7f267237b3ac55595a8f18170e4/run.go#L15 not the official ones.

- How can I install a different playwright version? The current is 1.7

**Comment by mxschmitt** [comment]
Hi @StarpTech,

thanks for reaching out.

Regarding the first issue, the language support on Playwright is based on a driver. This driver is then used like in the C#, Java or Python version to start the actual browsers. The communication is based on stdin/stdout. This driver is then installing the browsers [here](https://github.com/mxschmitt/playwright-go/blob/ca80d16b9f6030a0fa9738462c72ceea1b3c57e8/run.go#L46). And yes, the browsers are preinstalled in the official Docker image, but the driver isn't and unfortunately can't.

For the second one, yes it was until a few days where the driver was not yet hosted officially. There I migrated over to the official mirror, see [here](https://github.com/mxschmitt/playwright-go/blob/ca80d16b9f6030a0fa9738462c72ceea1b3c57e8/run.go#L204-L215) and #54.

The Go version and the driver version needs to be in sync. I will start later or in the next few days assigning logical version numbers to the repo as GitHub releases and tags, so it makes sense, as it's done in Python repository: https://github.com/microsoft/playwright-python/releases

Let me know if you have more questions and glad to hear that you like it. (There is a lot of potential in improving the option type handling and adding more tests, thats why I call it currently not yet production ready).

**Comment by StarpTech** [comment]
> Regarding the first issue, the language support on Playwright is based on a driver. This driver is then used like in the C#, Java or Python version to start the actual browsers. The communication is based on stdin/stdout. This driver is then installing the browsers here. And yes, the browsers are preinstalled in the official Docker image, but the driver isn't and unfortunately can't.

Does this mean we even have to download the browsers twice? This is esentially what I see when running my container

```
2020/12/22 16:46:12 Downloading driver...
2020/12/22 16:46:27 Downloaded driver successfully
2020/12/22 16:46:27 Downloading browsers...
Failed to process descriptor at /tmp/pw/node_modules/playwright
2020/12/22 16:46:28 Downloaded browsers successfully
```

> For the second one, yes it was until a few days where the driver was not yet hosted officially. There I migrated over to the official mirror, see here and #54.

> The Go version and the driver version needs to be in sync. I will start later or in the next few days assigning logical version numbers to the repo as GitHub releases and tags, so it makes sense, as it's done in Python repository: microsoft/playwright-python/releases

Great :+1:

**Comment by mxschmitt** [comment]
> Does this mean we even have to download the browsers twice? This is esentially what I see when running my container (based on playwright base image)

If the browser version will match with the browser installed in the Docker image, then they won't be downloaded again. This is blocked by the versioning, so its more clearer in the future.

I renamed the issue and we keep that open until I have the clear versioning in place, if you have more suggestions, feel free to open new issues, thanks!

**Comment by StarpTech** [comment]
@mxschmitt during testing it gets even worse. It downloads `.ms-playwright` for every subdir.

**Comment by mxschmitt** [comment]
> @mxschmitt during testing it gets even worse. It populates `.ms-playwright` for every subdir.

Thats currently the dilemma with my current approach. The driver needs to be downloaded a single time when it runs for the first time, currently during runtime which hangs in the beginning for a few seconds and its stored in the `$PWD/.mw-playwright`. If you want, you can create a small go binary which calls `playwright.Install()`.

So doing something like that would install the browsers in e.g. in Docker:

```go
package main

import (
	"log"

	"github.com/mxschmitt/playwright-go"
)

func main() {
	if err := playwright.Install(); err != nil {
		log.Printf("could not install: %v", err)
	}
}
```

Does this work for you approach or answer your question?

**Comment by StarpTech** [comment]
>  If you want, you can create a small go binary which calls playwright.Install().

I do not fully understand how it solves the issue because my $PWD is different for each test suite. I need a way to pass the location of the driver.

**Comment by mxschmitt** [comment]
> > If you want, you can create a small go binary which calls playwright.Install().
> 
> I do not fully understand how it will solve the issue because my $PWD is different for each test suite. I need a way to pass the location of the driver.

Not yet, we can add something like that. Let me come up with a proposal for it, will provide a PR for it soon.

**Comment by mxschmitt** [comment]
@StarpTech I rolled to Playwright v1.7.1 #66 but won't publish a release yet because I want to add missing functionality like video recording etc.

**Comment by mxschmitt** [comment]
Closed, I released version 0.171.0 which follows the same version schema as Playwright for Python etc. are using. It corresponds to version 1.7.1 of Playwright.
